examples/voting/src/lib.rs

```diff
@@ -0,0 +1,265 @@
+use std::collections::BTreeMap;
+
+use concordium_std::*;
+
+type VotingOption = String;
+type Vote = u32;
+type VoteCount = u32;
+
+#[derive(Serial, Deserial, SchemaType, Clone)]
+struct Description {
+    description_text: String,
+    options:          Vec<VotingOption>,
+}
+
+#[derive(Deserial, SchemaType)]
+struct VoteParameter {
+    vote: u32,
+}
+
+#[derive(Deserial, SchemaType)]
+struct InitParameter {
+    description: Description,
+    end_time:    Timestamp,
+}
+
+#[derive(Serial, Deserial, Clone, Eq, PartialEq, SchemaType)]
+struct Tally {
+    result:      BTreeMap<VotingOption, VoteCount>,
+    total_votes: VoteCount,
+}
+
+#[derive(Serial, Deserial, SchemaType)]
+struct VotingView {
+    description: Description,
+    tally:       Tally,
+    end_time:    Timestamp,
+}
+
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    description: Description,
+    ballots:     StateMap<AccountAddress, Vote, S>,
+    end_time:    Timestamp,
+}
+
+#[derive(Reject, Serial, PartialEq, Eq, Debug, SchemaType)]
+enum VotingError {
+    VotingFinished,
+    InvalidVoteIndex,
+    ParsingFailed,
+    ContractVoter,
+}
+
+impl From<ParseError> for VotingError {
+    fn from(_: ParseError) -> Self { VotingError::ParsingFailed }
+}
+
+type VotingResult<T> = Result<T, VotingError>;
+
+#[init(contract = "voting", parameter = "InitParameter")]
+fn init<S: HasStateApi>(
+    ctx: &impl HasInitContext,
+    state_builder: &mut StateBuilder<S>,
+) -> InitResult<State<S>> {
+    let param: InitParameter = ctx.parameter_cursor().get()?;
+    Ok(State {
+        description: param.description,
+        ballots:     state_builder.new_map(),
+        end_time:    param.end_time,
+    })
+}
+
+#[receive(
+    contract = "voting",
+    name = "vote",
+    mutable,
+    parameter = "VoteParameter",
+    error = "VotingError"
+)]
+fn vote<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+) -> VotingResult<()> {
+    // Check that the election hasn't finished yet.
+    if ctx.metadata().slot_time() > host.state().end_time {
+        return Err(VotingError::VotingFinished);
+    }
+    // Ensure that the sender is an account.
+    let acc = match ctx.sender() {
+        Address::Account(acc) => acc,
+        Address::Contract(_) => return Err(VotingError::ContractVoter),
+    };
+    let param: VoteParameter = ctx.parameter_cursor().get()?;
+
+    let new_vote = param.vote;
+
+    // Check that vote is in range
+    ensure!(
+        (new_vote as usize) < host.state().description.options.len(),
```

Comment:
We are loading in a potentially large vec here, just for the length.
We should probably just store the length separately in the state.

Reply:
`num_options` is now a state variable.